Add multiple projects per repo support

SDBowen opened the pull request:
Got this working on dev, but probably needs some more testing.

In the local `boards` folder, I created a single repo folder with a single kitspace.yaml in the root folder, and two project folders:

![Screenshot from 2019-05-19 21-04-06](https://user-images.githubusercontent.com/22715037/57986800-be121b80-7a79-11e9-8f67-301d34963153.png)

Here is the kitspace.yaml

![Screenshot from 2019-05-19 21-06-00](https://user-images.githubusercontent.com/22715037/57986818-eef25080-7a79-11e9-93a6-7f0c5db37110.png)

The build:

![Screenshot from 2019-05-19 21-06-51](https://user-images.githubusercontent.com/22715037/57986829-0f220f80-7a7a-11e9-86b4-31bba2adef73.png)
![Screenshot from 2019-05-19 21-07-28](https://user-images.githubusercontent.com/22715037/57986839-35e04600-7a7a-11e9-9500-e79b2ab0cb33.png)
![Screenshot from 2019-05-19 21-07-46](https://user-images.githubusercontent.com/22715037/57986841-3aa4fa00-7a7a-11e9-89d1-cd09c225ffae.png)

kasbah commented:
Looking good! I fixed the formatting issues on master and [rebased your commit for a clearer diff](https://github.com/kitspace/kitspace/commit/4a39200204e6e8d8a2c36f73e3ab78aa32f4555a) feel free to reset this branch to that commit. 

[Preview is here in case you didn't know](http://multiple-projects-per-repo.preview.kitspace.org/). Seems like it hasn't broken anything. I'll set up an example project here this week: https://github.com/kitspace-forks/hoverboard-breakout

I think we should enforce that the name of the project is the `path` instead of adding and extra field. That avoids some confusion around what's what. I also think that overrides to the `gerber` and `bom` paths should be relative to the git root rather than the `path`.

SDBowen commented:
Agree on getting rid of the `path` field.

I will look at changing the `gerber` and `bom` paths. Will try and do it soon while it is still fresh in my mind!

SDBowen commented:
The bom override is now relative to the repo root and seems to be working.

Having trouble with the gerber override. Without a `gerbers` entry in the yaml file, it works fine. When trying to specify a `gerbers` path, all files get filtered.

kasbah commented on tasks/page/processGerbers.js:
```diff
@@ -11,17 +11,35 @@ const gerberFiles = require('../../src/gerber_files')
+    let info = file == null ? {} : yaml.safeLoad(file)
     const files = globule
       .find(`${folder}/**/*`)
       .map(p => path.relative(folder, p))
```
I think this should be searching `repoRootPath` and be relative to it. Maybe this solves your issue?

SDBowen replied:
Thanks @kasbah, that was the issue.

kasbah commented:
I couldn't properly debug because I didn't have your example. I made an example project. Could you add https://github.com/kitspace-forks/hoverboard-breakout to `boards.txt` and `dev_registry.json` please and test with that?

SDBowen commented:
I had to format the ```BOM.csv``` headers in the ```bobbycar_breakout``` board. After that, the build was successful.

kasbah commented:
Ok, you should update [the repo](https://github.com/kitspace-forks/hoverboard-breakout) with the BOM.csv changes then (and make sure the right version is in dev_registry.json). 

Not sure why CI hasn't triggered yet. :thinking:

kasbah commented:
I think we need to clean up how gerbers are handled. At some point I made it search through the repo to look for what looks like gerber files but that may have been a mistake. They were supposed to always be in a `gerbers` directory in the root or overridden using the `gerbers: ` field in the yaml. 

The way it is currently, with multi projects this will cause unexpected behavior with it picking up random gerber files even if you have clearly put them into the the default folder names and specified that in the manifest.  E.g. this will result in the wrong gerbers getting picked up for either project_1 or project_2 which will be surprising. 

```yaml
multi:
    project_1:
        summary: project 1
    project_2:
        summary: project 2
```

```
.
├── kitspace.yaml
├── project_1
│   └── gerbers
│       ├── project_1.cmp
│       ├── project_1.drd
│       ├── project_1.dri
│       ├── project_1.gko
│       ├── project_1.gpi
│       ├── project_1.gto
│       ├── project_1.plc
│       ├── project_1.sol
│       ├── project_1.stc
│       └── project_1.sts
└── project_2
    └── gerbers
        ├── project_2.cmp
        ├── project_2.drd
        ├── project_2.dri
        ├── project_2.gko
        ├── project_2.gpi
        ├── project_2.gto
        ├── project_2.plc
        ├── project_2.sol
        ├── project_2.stc
        └── project_2.sts
```

kasbah commented:
Looks like Travis CI is having issues. https://www.traviscistatus.com/

SDBowen commented:
Should the change on gerbers handling be done on a different branch? 

Instead of searching the repo, we will just look in a `gerbers` directory, unless a `bom` override exists. Is this the approach you are thinking?  Do all projects on Kitspace follow this currently?

Good to see this build didn't break the CI.

kasbah commented:
Looks like CI is running again and needs the fix for the BOM. https://travis-ci.org/kitspace/kitspace/builds/535886482#L3525

There may be a few projects that are making use of the search currently. We could also continue to search but only search the right paths on multi projects. That's probably the better way to do it but may be a bit more complicated.

SDBowen commented:
Spoke too soon, looks like the build did fail...

Trying to update the `kitspace-forks/hoverboard-breakout` repo I get the following error:

![Screenshot from 2019-05-22 20-47-24](https://user-images.githubusercontent.com/22715037/58200597-6b33a080-7cd3-11e9-8781-36559e101c19.png)

kasbah commented:
I sent you an invite to that organisation, maybe you need to accept that?

SDBowen commented:
Updated the gerber files search to look only at the project folders for multi repos. :+1: 

Also changed `board_card.jsx`; grabs the last folder in the path as the card name:

![Screenshot from 2019-05-24 20-48-13](https://user-images.githubusercontent.com/22715037/58350398-0eb4ba80-7e66-11e9-9849-78089c2f1953.png)

I was still having trouble with `kitspace-forks/hoverboard-breakout`, but after forking the repo, the PR went through. After that PR is approved, I will update the `dev_registry.json` on this branch, and Travis CI will pass.

kasbah commented on src/index/board_card.jsx:
```diff
@@ -70,7 +70,7 @@ let BoardCard = React.createClass({
             {this.props.data.id
               .split('/')
-              .slice(2)
+              .slice(-1)
```
Does it get too long otherwise?

SDBowen replied:
It did for that test report. 

It could also be handled by resizing the text, if you want to preserve the title with the project path.

kasbah replied:
Makes sense, let's not go down a style tweaks rabbit hole as it seems fine like this.

kasbah commented on tasks/page/processGerbers.js:
```diff
@@ -29,17 +28,21 @@ if (require.main !== module) {
       for (let project in info.multi) {
         if (project === projectPath) {
           info = info.multi[project]
+          gerberPath = `${repoRootPath}/${projectPath}/**/*`
```
We should probably be using `path.join` more to avoid mistakes with extra `/`.

SDBowen replied:
Much cleaner then the template literals.

kasbah commented on tasks/page/processGerbers.js:
```diff
@@ -29,17 +28,21 @@ if (require.main !== module) {
+
     const gerbers = gerberFiles(files, info.gerbers).map(p =>
       path.join(repoRootPath, p)
     )
```
I think this way it wouldn't allow gerbers to be kept outside of the `projectPath` but we should give that flexibility to people that want to organize their repo in a different way. E.g.

```yaml
multi:
    project_1:
        gerbers: gerbers_1/
    project_2:
        gerbers: gerbers_2/
```

```
.
├── gerbers_1
│   ├── gerber_1.gbr
│   └── gerber_2.gbr
├── gerbers_2
│   ├── gerber_1.gbr
│   └── gerber_2.gbr
├── kitspace.yaml
├── project_1
│   ├── cad_file.cad
│   └── README.md
└── project_2
    ├── cad_file.cad
    └── README.md
```

SDBowen replied:
Agreed. Think I got this figured out.

kasbah commented:
Preview is up! http://multiple-projects-per-repo.preview.kitspace.org/

kasbah commented:
https://twitter.com/kitspaceorg/status/1132012184794533888

SDBowen commented:
Now that it is on Twitter, guess I have to see the PR through!

kasbah commented:
Looks like it's done except for documentation?

SDBowen commented:
Updated the readme for the Kitspace repo. 

The 'Submit a project' preview will need some changes to work correctly with this PR change. I haven't dove into that too much yet :upside_down_face:

kasbah commented on README.md:
```diff
@@ -116,6 +117,31 @@ bom: manufacture/advanced-example-BOM.tsv
+        summary: First project in a repository.
+        color: blue
+        site: https://example-one.com
+    Project_two:
```
`s/Project/project/`

SDBowen replied:
Can you expand on that? 

I see I capitalized `project_two` on accident. Add one more folder layer?

kasbah replied:
no no, just that capitalization.

SDBowen replied:
Cool, got it fixed :+1:

kasbah commented:
I just noticed that the title is short on the cards and then the full one on the page which is probably a little confusing.

![image](https://user-images.githubusercontent.com/206854/58704796-b5381880-839c-11e9-88e9-dda56fe1657b.png)
![image](https://user-images.githubusercontent.com/206854/58704808-c2550780-839c-11e9-99e1-9e17ea3ff3ec.png)

I think the best solution would be to add it to sub-title on both card and page i.e. `github.com / kitspace-forks / hoverboard-breakout`

kasbah commented:
Though I am not sure it's the right approach either, as that field seems cleaner as always the "user".  I think it should be consistent from card to page though. I think actually just remove it from the page as well for now since we don't have space on the card.

SDBowen commented:
Updated the project page titles. 

Should subtitles (card and page) be updated to show the full path, `github.com / kitspace-forks / hoverboard-breakout` instead of `github.com / kitspace-forks`?

kasbah commented:
Thanks, nah, I was thinking about it but it's as is for now I think.

Am I ok to force push to this branch? I need to correct some commits I made.

kasbah commented:
I'll just add a new commit. Need to get out the habit of force pushing.
